sys/trickle/trickle.c

```diff
@@ -30,63 +27,61 @@ void trickle_callback(trickle_t *trickle)
     if ((trickle->c < trickle->k) || (trickle->k == 0)) {
         (*trickle->callback.func)(trickle->callback.args);
     }
+
+    trickle_interval(trickle);
 }
 
 void trickle_interval(trickle_t *trickle)
 {
-    uint32_t max_interval;
+    uint32_t old_interval = trickle->I;
+    uint32_t max_interval = trickle->Imin << trickle->Imax;
+    uint32_t diff = old_interval - trickle->t;
 
-    trickle->I = trickle->I * 2;
-    max_interval = trickle->Imin << trickle->Imax;
+    trickle->I *= 2;
 
     if ((trickle->I == 0) || (trickle->I > max_interval)) {
         trickle->I = max_interval;
     }
 
-    DEBUG("trickle: I == %" PRIu32 "\n", trickle->I);
+    DEBUG("trickle: I == %" PRIu32 ", diff == %" PRIu32 "\n", trickle->I, diff);
 
     trickle->c = 0;
-    trickle->t = (trickle->I / 2) + random_uint32_range(0, (trickle->I / 2) + 1);
+    trickle->t = random_uint32_range(old_interval, trickle->I);
 
-    trickle->msg_callback_time = trickle->t * MS_PER_SEC;
-    xtimer_set_msg64(&trickle->msg_callback_timer, trickle->msg_callback_time,
-                     &trickle->msg_callback, trickle->pid);
+    trickle->msg_time = (trickle->t + diff) * MS_PER_SEC;
+    xtimer_set_msg64(&trickle->msg_timer, trickle->msg_time, &trickle->msg,
+                     trickle->pid);
 
-    trickle->msg_interval_time = trickle->I * MS_PER_SEC;
-    xtimer_set_msg64(&trickle->msg_interval_timer, trickle->msg_interval_time,
-                     &trickle->msg_interval, trickle->pid);
 }
 
 void trickle_reset_timer(trickle_t *trickle)
 {
     trickle_stop(trickle);
-    trickle_start(trickle->pid, trickle, trickle->msg_interval.type, trickle->msg_callback.type,
-                  trickle->Imin, trickle->Imax, trickle->k);
+    trickle_start(trickle->pid, trickle, trickle->msg.type, trickle->Imin,
+                  trickle->Imax, trickle->k);
 }
 
-void trickle_start(kernel_pid_t pid, trickle_t *trickle, uint16_t interval_msg_type,
-                   uint16_t callback_msg_type, uint32_t Imin, uint8_t Imax, uint8_t k)
+void trickle_start(kernel_pid_t pid, trickle_t *trickle, uint16_t msg_type,
+                   uint32_t Imin, uint8_t Imax, uint8_t k)
 {
     trickle->pid = pid;
 
     trickle->c = 0;
     trickle->k = k;
     trickle->Imin = Imin;
     trickle->Imax = Imax;
-    trickle->I = trickle->Imin + random_uint32_range(0, 4 * trickle->Imin);
+    trickle->I = trickle->t = random_uint32_range(trickle->Imin,
+                                                  4 * trickle->Imin);
```

Comment:
actually it's not in the RFC; the previous implementer had this and I adopted it. There's no guideline on how to choose the initial `I` value for trickle. The only constraint from the RFC is: `I` must be a value between `Imin` and `Imax`. Making the right hand side configurable on compile time might be a way to go though.